Ocean/sediment transport design doc

caozd999 opened the pull request:
This PR adds the prognostic sediment transport design document. It would also be used for discussions about where and how
-  to define the sediment transport related variables (have been/will be listed in the design document) 
-  to implement each development requirement  

This design document will keep updated along with the code development work.

xylar commented:
@caozd999, I've added @mark-petersen and myself as reviewers but neither of us knows anything about sediment transport.  Who else should review for the science side, not just technical details?

xylar commented:
@caozd999, thanks for submitting this at my request!  I'm going to do a first pass through the design doc just for formatting issues related to RestructuredTest and sphinx.  I hope that's okay.  I probably won't get to that until tomorrow, though.  Too many meetings today.

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+development work in `Warner et al., 2005 <https://doi.org/10.1016/j.cageo.2008.02.012>`_ and
+the `COAWST code (USGS version) <https://github.com/jcwarner-usgs/COAWST>`_.
+The main model components include
+1) sediment classes;
+2) bedload sediment transport;
+3) suspended sediment transport; and
+4) morphological change.
```
If you want this to come out formatted like a numbered list, you need blank lines:
```suggestion

1) sediment classes;

2) bedload sediment transport;

3) suspended sediment transport; and

4) morphological change.

```

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+This approach requires the knowledge of the near-bottom reference SSC,
+which can be computed by one of the following two methods:
+
+- `Zyserman and Fredsoe, 1994 <https://ascelibrary.org/doi/pdf/10.1061/%28ASCE%290733-9429%281994%29120%3A9%281021%29>`_
+- `Goldstein et al., 2014 <https://esurf.copernicus.org/articles/2/67/2014/esurf-2-67-2014-discussion.html>`_
```
```suggestion
- `Zyserman and Fredsoe, 1994 <https://ascelibrary.org/doi/pdf/10.1061/%28ASCE%290733-9429%281994%29120%3A9%281021%29>`_

- `Goldstein et al., 2014 <https://esurf.copernicus.org/articles/2/67/2014/esurf-2-67-2014-discussion.html>`_
```

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+the fluid characteristics (density and viscosity), and well as the fluid turbulence velocity.
+Two sediment settling veloity formulations are selected to implement into MPAS-Ocean, including
+
+- `Soulsby, 1997 <https://www.icevirtuallibrary.com/doi/abs/10.1680/doms.25844.fm>`_
+- `Cheng, 1997 <https://doi.org/10.1061/(ASCE)0733-9429(1997)123:2(149)>`_
```
```suggestion
- `Soulsby, 1997 <https://www.icevirtuallibrary.com/doi/abs/10.1680/doms.25844.fm>`_

- `Cheng, 1997 <https://doi.org/10.1061/(ASCE)0733-9429(1997)123:2(149)>`_
```

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+The main difference of the two formulations is the computation of :math:`\Phi`, which is a function of another
+dimensionless parameter called Shields parameter :math:`\theta`
+
+.. math::
+	\theta = \frac{\tau}{g(s-1)D_{50}}
```
```suggestion
.. math::

	\theta = \frac{\tau}{g(s-1)D_{50}}
```

xylar replied:
You're also using tabs for indenting in a few places.  The document should not have any tabs, just spaces.

caozd999 replied:
Thanks. Is there any specific number of spaces I should place for the indenting format?

xylar replied:
I follow the python specification of 4 spaces but I don't think RST/sphinx cares.

xylar replied:
I looked into this further.  The preference is for 3 spaces: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html#unordered-list-inside-ordered-list

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+
+The Meyer-Peter-Mueller and Soulsby-Dammgaard share the same expression of the bedload transport rate
+
+.. math::
+    q_b = \Phi \sqrt{(s-1)g D_{50}^3}\rho_s
```
```suggestion
.. math::

    q_b = \Phi \sqrt{(s-1)g D_{50}^3}\rho_s
```

xylar replied:
Same in many other places.

caozd999 replied:
I will work through the document to fix them.

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+- A logical variable to switch ON/OFF the bedload transport: :code:`config_sediment_bedload`
+
+- A character variable to define the bedload transport method: :code:`config_sediment_bedload_formulation`
+  And there are three options for it:
+	* :code:`Engelund-Hansen`
+	* :code:`Meyer-Peter-Mueller`
+	* :code:`Soulsby-Damgaard`
```
```suggestion
  And there are three options for it:

        * :code:`Engelund-Hansen`

	* :code:`Meyer-Peter-Mueller`

	* :code:`Soulsby-Damgaard`
```

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,476 @@
+
+
+Bedload transport also requires the definitions of the following variables:
+
+- Pre-defined sediment-related variables
+	* sediment grain diameter: :code:`sedimentGrainDiameter(nSedimentClasses, nCells)`
+	* sediment grain density: :code:`sedimentGrainDensity(nSedimentClasses, nCells)`
+	* sediment class fraction on bed layer: :code:`sedimentClassFraction(nSedimentClasses)`
+	* bedload transport coefficient: :code:`sedimentBedloadCoefficient(nSedimentClasses)`
+- Other variables
+	* bedload transport rate: :code:`sedimentBedloadTransportRate(nCells)`
```
```suggestion

- Pre-defined sediment-related variables

  * sediment grain diameter: :code:`sedimentGrainDiameter(nSedimentClasses, nCells)`

  * sediment grain density: :code:`sedimentGrainDensity(nSedimentClasses, nCells)`

  * sediment class fraction on bed layer: :code:`sedimentClassFraction(nSedimentClasses)`

  * bedload transport coefficient: :code:`sedimentBedloadCoefficient(nSedimentClasses)`

- Other variables

  * bedload transport rate: :code:`sedimentBedloadTransportRate(nCells)`
```

xylar replied:
I guess you get the idea.  Many of these lists need this kind of spacing.

xylar commented:
@caozd999, I think that's all my tech feedback for now.

caozd999 commented:
> @caozd999, I've added @mark-petersen and myself as reviewers but neither of us knows anything about sediment transport. Who else should review for the science side, not just technical details?

I think LeAnn should be added here -- she is team member of this task.

xylar commented:
@lconlon, if you would review the design doc, that would be great!

@mark-petersen and @lconlon, to build the docs locally, you need to install [Miniconda3](https://docs.conda.io/en/latest/miniconda.html), add the [conda-forge](https://conda-forge.org/#about) channel, and run:
```
conda create -y -n mpas_docs python=3.8 sphinx mock sphinx_rtd_theme
conda activate mpas_docs
export DOCS_VERSION="test"
cd docs
make html
```
Then, open `_build/html/index.html` in your browser.  @caozd999 has trouble building on his Mac.  Hopefully, this has been solved but maybe use Grizzly as a backup.

caozd999 commented:
> @lconlon, if you would review the design doc, that would be great!
> 
> @mark-petersen and @lconlon, to build the docs locally, you need to install [Miniconda3](https://docs.conda.io/en/latest/miniconda.html), add the [conda-forge](https://conda-forge.org/#about) channel, and run:
> 
> ```
> conda create -y -n mpas_docs python=3.8 sphinx mock sphinx_rtd_theme
> conda activate mpas_docs
> export DOCS_VERSION="test"
> cd docs
> make html
> ```
> 
> Then, open `_build/html/index.html` in your browser. @caozd999 has trouble building on his Mac. Hopefully, this has been solved but maybe use Grizzly as a backup.

Thanks @xylar. After I replaced tabs by spaces and added spaces between bullet lines, I got the right document format on html.  And I found it much faster to build html locally when I turn LANL VPN off.

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+We provde two options to compute :math:`C_{ref}` and :math:`z_{ref}`:
+
+- *Zyserman and Fredsoe, 1994*
+
+.. math::
+
+      C_{ref} = \frac{0.331(\theta - 0.045)^{1.75}}{1+0.72(\theta-0.045)^{1.75}}
+
+and :math:`z_{ref}=2D_{50}`. Again, :math:`\theta` is the Shields parameter and
+:math:`D_{50}` is the sediment median grain diameter.
+
+- *Goldstein et al., 2014*
+
+.. math::
+
+      C_{ref} = \left[ \frac{0.328U_b}{0.0688+1000D_{50}}\right]^2
+
+at height :math:`z_{ref}=0.01 m`, where :math:`U_b` is the magnitude of the bottom velocity (:math:`ms^{-1}`)
+
```
Based on this link: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html#code-images-and-other-content-inside-lists

```suggestion
- *Zyserman and Fredsoe, 1994*

  .. math::

      C_{ref} = \frac{0.331(\theta - 0.045)^{1.75}}{1+0.72(\theta-0.045)^{1.75}}

  and :math:`z_{ref}=2D_{50}`. Again, :math:`\theta` is the Shields parameter and
:math:`D_{50}` is the sediment median grain diameter.

- *Goldstein et al., 2014*

  .. math::

      C_{ref} = \left[ \frac{0.328U_b}{0.0688+1000D_{50}}\right]^2

  at height :math:`z_{ref}=0.01 m`, where :math:`U_b` is the magnitude of the bottom velocity (:math:`ms^{-1}`)

```

xylar replied:
That is, everything but the "bullet" lines need to be indented 2 spaces.

caozd999 replied:
Got it.

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+      w_s = \frac{\nu}{D_{50}}\left[\sqrt{ (10.36^2+1.049D_*^3)}-10.36\right]
+
+where :math:`\nu` is the kinematic viscosity of water (:math:`m^2s^{-1}`)
```
This needs to be indented, too:
```suggestion
  where :math:`\nu` is the kinematic viscosity of water (:math:`m^2s^{-1}`)
```

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+      w_s = \frac{\nu}{D_{50}} (\sqrt{25+1.2D_*^2}-5)^{1.5}
+
+in which :math:`D_* = D_{50}\left[\frac{(s-1)g}{\nu ^2}\right]^{1/3}` is dimensionless grain size.
```
```suggestion
  in which :math:`D_* = D_{50}\left[\frac{(s-1)g}{\nu ^2}\right]^{1/3}` is dimensionless grain size.
```

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+- Pre-defined sediment-related variables
+
+  * sediment settling velocity: :code:`sedimentSettlingVelocity(nSedimentClasses)`
+
+  * sediment critical erosion shear: :code:`sedimentCriticalErosionShear(nSedimentClasses)`
+
+  * sediment erodibility constant: :code:`sedimentErodibilityConstant(nSedimentClasses)`
+
+  * sediment bed porosity: :code:`sedimentBedPorosity(nSedimentClasses)`
+
+- Other variables
+
+  * sediment suspended concentration: :code:`sedimentSuspendedConcentration(nSedimentClasses, nCell)`
```
Again, a corresponding variable name from Algorithm Design: Suspended sediment transport would be helpful for each variable here.

The bottom velocity might warrant some discussion.  For one, the velocity magnitude isn't necessarily available at cell centers (but it can be computed from the kinetic energy).  For another, you may want to average over more than one layer to smooth out jumps depending on what vertical coordinates and bathymetries this capability supports (see my suggested requirement related to this.)

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+Contributors: Zhendong Cao
+
+A logical variable need to be defined to swith ON/OFF the morphological change: :code:`config_morphological_change`.
```
```suggestion
A logical variable need to be defined to switch ON/OFF the morphological change: :code:`config_morphological_change_enabled`.
```

caozd999 replied:
corrected.

xylar commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+  * sediment suspended concentration: :code:`sedimentSuspendedConcentration(nSedimentClasses, nCell)`
+
+Implementation: Morphological change
```
This may not seem like it but it might be the hardest change to make.  You may want to meet us with me, @dengwirda, @cbegeman and @stephenprice to discuss this because it may relate to work we are doing for ice-shelf cavities.  Currently, `bottomDepth` is the (positive-down) bathymetry in MPAS-Ocean, but it isn't allowed to change in time.  It is also intimately tied to the z* vertical coordinate that is currently the default and is implicitly assumed in at least a few places in the code that @cbegeman has discussed with us (related to the computation of vertical velocities between layers).

A change to the model to allow `bottomDepth` to change with time could be a pretty big project and could potentially be a bit disruptive.  @matthewhoffman may also be interested in this work, since it also relates to his early career project.  In sort, this might be a big enough project to need a design document all on its own with input from a lot of different people.

cbegeman replied:
@caozd999, @lconlon and I have had a conversation about implementing morphological change to the bathymetry. Correct me if I'm wrong, but I think our conclusion was that this development wasn't strictly necessary for ICoM's goals and that it likely wouldn't be feasible in the [current development timeframe](https://icom.atlassian.net/wiki/spaces/ICOM/pages/101711907/Overview). I'd suggest removing it from this design doc and if we find ourselves ahead of the timeline or if this development seems necessary for ICoM then we create a new design document for morphological change alone.

lconlon replied:
I'm trying to remember that conversation now....would erosion and deposition still work correctly if we don't change the bed elevation in the model? I remember that we decided that changing coastline morphology was out of the scope of this project, but I can't remember where we landed on bed elevation.

caozd999 replied:
@cbegeman you are right, we don't have to implement the morphological change component for ICoM now; I put it in this doc for a more general model component introduction.  As @xylar and @cbegeman suggested, I will remove morphological change from the design doc, and we can prepare a single design doc for when necessary. But it is good to know that this would influence/interest other people if we tend to do it; seems to me a necessary model development direction.

caozd999 replied:
@lconlon we would just assume the erosion and deposition are not strong enough to significantly change either the bed layer sediment mass or the sediment layer thickness. So it would work. And yes you are right - we had a discussion before and the coastline morphology won't be in the scope either.

xylar reviewed (commented):
@caozd999, this is a great document!  I've made a bunch of comments.  Probably the second to last on Implementation: Morphological change is the most critical for us to talk about.  Changing the bathymetry is likely to be a pretty big project!

lconlon reviewed (commented):
I think the science here looks fine and is in line with what we've already talked about. I don't have any specific comments on it but I'm happy to chat about specifics if needed. Thanks for putting this together!

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+The number of sediment classes will be defined manually and considered as a dimension of sediment-related variables.
+
+Algorithm Design: Bedload sediment transport
```
I plan to implement both and let user choose.

caozd999 commented:
> I think the science here looks fine and is in line with what we've already talked about. I don't have any specific comments on it but I'm happy to chat about specifics if needed. Thanks for putting this together!

Thanks @lconlon. Within the ICoM scope, I think we'd better remove the morphological change part, but we do consider the erosion and deposition when we solve SSC. Does this sound good to you?

lconlon commented:
> > I think the science here looks fine and is in line with what we've already talked about. I don't have any specific comments on it but I'm happy to chat about specifics if needed. Thanks for putting this together!
> 
> Thanks @lconlon. Within the ICoM scope, I think we'd better to remove the morphological change part, but we do consider the erosion and deposition when we solve SSC. Does this sound good to you?

That sounds good. Thanks for all your work on this!

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+time step.
+
+The sucessful implementation will enable MPAS-Ocean to simulate the sediment transport (especially
+the computation of suspended sediment cocentration (SSC)) with given initial and boundary conditions,
```
corrected.

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+its own source/sink terms, which need to resolve the deposition and resuspension at each
+time step.
+
+The sucessful implementation will enable MPAS-Ocean to simulate the sediment transport (especially
```
corrected.

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+Bedload and suspended sediment transport are the two main transport forms considered in this development.
+Morphological change provides the trigger for interactions between different model components.
+
+The primary implimentation challenges lie in the suspended sediment transport component.
```
corrected.

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+Morphological change provides the trigger for interactions between different model components.
+
+The primary implimentation challenges lie in the suspended sediment transport component.
+It requires to solve the three-dimensional scalar equation for suspended sediment with
```
corrected.

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+	dim name='nSedimentClasses' unit='unitless'
+	    description='the number of sediment classes considered in the model'
+
+Implementation: Bedload sediment transport
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I think these are intermediate variables that maybe don't need to be defined externally.

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+                    description="Controls if sediment bedload transport is computed."
+                    possible_values=".true. or .false."
+        />
+        <nml_option name="config_sediment_suspended_formulation" type="chracter" default_value="None" units="unitless"
+                    description="Select the sediment bedload transport formulation"
+                    possible_values=" 'None', 'Zyserman-Fredsoe', 'Goldstein'"
+	/>
```
Yes, that's right. I will add `config_sediment_settling_formulation`in the registry file.

caozd999 commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,519 @@
+
+Bedload transport also requires the definitions of the following variables:
+
+- Pre-defined sediment-related variables
+	
+  * sediment grain diameter: :code:`sedimentGrainDiameter(nSedimentClasses, nCells)`
+
+  * sediment grain density: :code:`sedimentGrainDensity(nSedimentClasses, nCells)`
+
+  * sediment class fraction on bed layer: :code:`sedimentClassFraction(nSedimentClasses)`
+
+  * bedload transport coefficient: :code:`sedimentBedloadCoefficient(nSedimentClasses)`
+
+- Other variables
+	
+  * bedload transport rate: :code:`sedimentBedloadTransportRate(nCells)`
```
`\rho_w` should be from the lowest ocean layer. I will remove the wave effects for just now.

xylar commented:
@caozd999, how are you feeling about this one?  Is it at a stage where you're ready to merge it?  You can always edit it later.

We're in the process of merging PRs wherever we can to get ready for some pretty big changes to MPAS-Ocean.  Happy to chat about it more on Slack.

caozd999 commented:
> @caozd999, how are you feeling about this one? Is it at a stage where you're ready to merge it? You can always edit it later.
> 
> We're in the process of merging PRs wherever we can to get ready for some pretty big changes to MPAS-Ocean. Happy to chat about it more on Slack.

Thanks @xylar. I think this version is ready to merge, and we can edit it later if necessary.

mark-petersen commented:
@xylar @lconlon @cbegeman if any of you are done reviewing, please approve. This PR is good for an 'off' day when Jon is out.

cbegeman commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,415 @@
+:math:`u_*` is the shear friction velocity (:math:`m\,s^{-1}`).
+The shear stress :math:`\tau = C_d \overline{U}^2={u_*}^2`, so :math:`u_*={C_d}^{0.5} \overline{U}`.
+
+We provde two options to compute :math:`C_{ref}` and :math:`z_{ref}`:
```
```suggestion
We provide two options to compute :math:`C_{ref}` and :math:`z_{ref}`:
```

mark-petersen replied:
done

cbegeman commented on docs/ocean/design_docs/prognostic-sediment-transport.rst:
```diff
@@ -0,0 +1,415 @@
+
+The sediment bedload transport requires the pre-definitions of the following namelist options:
+
+- A logical variable to switch ON/OFF the bedload transport: :code:`config_sediment_bedload_enabled`
```
```suggestion
- A logical variable to switch ON/OFF the bedload transport: :code:`config_sediment_bedload_enable`
```
For consistency with some of the other config variables

mark-petersen replied:
I changed all these, thanks. @caozd999 all the flags use `enable` at the end without the d

caozd999 replied:
@mark-petersen Many thanks Mark.

cbegeman commented:
@caozd999 and @mark-petersen I'm happy to have this merged as-is.
